platform/darwin/src/MGLOfflineTask.mm

```diff
@@ -0,0 +1,198 @@
+             @"on this instance of MGLOfflineTask, rendering it invalid. It is an " \
+             @"error to send any message to this task."]; \
+        } \
+    } while (NO);
+
+class MBGLOfflineRegionObserver : public mbgl::OfflineRegionObserver {
+public:
+    MBGLOfflineRegionObserver(MGLOfflineTask *offlineTask_) : offlineTask(offlineTask_) {}
+    
+    void statusChanged(mbgl::OfflineRegionStatus status) override;
+    void responseError(mbgl::Response::Error error) override;
+    void mapboxTileCountLimitExceeded(uint64_t limit) override;
+    
+private:
+    __weak MGLOfflineTask *offlineTask = nullptr;
+};
+
+@interface MGLOfflineTask ()
+
+@property (nonatomic, readwrite) mbgl::OfflineRegion *mbglOfflineRegion;
+@property (nonatomic, readwrite) MGLOfflineTaskState state;
+@property (nonatomic, readwrite) MGLOfflineTaskProgress progress;
+
+@end
+
+@implementation MGLOfflineTask
+
+- (instancetype)init {
+    [NSException raise:@"Method unavailable"
+                format:
+     @"-[MGLOfflineTask init] is unavailable. "
+     @"Use +[MGLOfflineStorage addTaskForRegion:withContext:completionHandler:] instead."];
+    return nil;
+}
+
+- (instancetype)initWithMBGLRegion:(mbgl::OfflineRegion *)region {
+    if (self = [super init]) {
+        _mbglOfflineRegion = region;
+        _state = MGLOfflineTaskStateUnknown;
+        
+        mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+        MBGLOfflineRegionObserver *mbglObserver = new MBGLOfflineRegionObserver(self);
+        mbglFileSource->setOfflineRegionObserver(*_mbglOfflineRegion, std::make_unique<MBGLOfflineRegionObserver>(*mbglObserver));
+    }
+    return self;
+}
+
+- (void)dealloc {
+    mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+    mbglFileSource->setOfflineRegionObserver(*_mbglOfflineRegion, nullptr);
+}
+
+- (id <MGLOfflineRegion>)region {
+    MGLAssertOfflineTaskIsValid();
+    
+    const mbgl::OfflineRegionDefinition &regionDefinition = _mbglOfflineRegion->getDefinition();
+    NSAssert([MGLTilePyramidOfflineRegion conformsToProtocol:@protocol(MGLOfflineRegion_Private)], @"MGLTilePyramidOfflineRegion should conform to MGLOfflineRegion_Private.");
+    return [(id <MGLOfflineRegion_Private>)[MGLTilePyramidOfflineRegion alloc] initWithOfflineRegionDefinition:regionDefinition];
+}
+
+- (NSData *)context {
+    MGLAssertOfflineTaskIsValid();
+    
+    const mbgl::OfflineRegionMetadata &metadata = _mbglOfflineRegion->getMetadata();
+    return [NSData dataWithBytes:&metadata[0] length:metadata.size()];
+}
+
+- (void)resume {
+    MGLAssertOfflineTaskIsValid();
+    
+    mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+    mbglFileSource->setOfflineRegionDownloadState(*_mbglOfflineRegion, mbgl::OfflineRegionDownloadState::Active);
+    self.state = MGLOfflineTaskStateActive;
```

Comment:
I'm concerned about `MGLOfflineTask` having its own `state` ivar that's not necessarily in sync with the underlying `mbgl::OfflineRegion`'s state.

One way the two will get out of sync is a consequence of the fact that `setOfflineRegionDownloadState` is asynchronous. It's possible to receive a few `statusChanged` notifications with a `downloadState` of `Active` even after calling `setOfflineRegionDownloadState(Inactive)`, prior the async state change being processed. Once it is processed, the implementation guarantees to send you a final `statusChanged` notification with a `downloadState` of `Inactive`. In this scenario, the current iOS implementation will flip `self.state` from `Active` ⇢ `Inactive` ⇢ `Active` ⇢ `Inactive`.

I suggest making `self.state` entirely determined by the value last received in `statusChanged`, with two exceptions:

* The initial state is always `MGLOfflineTaskStateUnknown`
* Once the state goes to `MGLOfflineTaskStateInvalid`, further state changes from `statusChanged` are ignored